{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Modules import generate_bed_files\n",
    "from Modules import get_gene_exons\n",
    "from Modules import exon_dist\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_target_genes():\n",
    "    df = pd.read_csv('Datasets/Lab_Pvalb-tpm.tsv', sep='\\t', header=0, index_col=0, usecols=[0,1,2])\n",
    "    df = df.loc[df.index.str.startswith('Hb'),:]\n",
    "    \n",
    "    return df.index.tolist()\n",
    "\n",
    "def generate_references():\n",
    "    genes = get_target_genes()\n",
    "    df = get_gene_exons.generate_genome(genelist=genes)\n",
    "    for gene in genes:\n",
    "        get_gene_exons.create_gene_reference_file(df, gene)\n",
    "    \n",
    "    return\n",
    "\n",
    "def get_genes_data():\n",
    "    genes = get_target_genes()\n",
    "    df = get_gene_exons.get_genes_data(genes)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_targets():\n",
    "    df = get_genes_data()\n",
    "    targets = generate_bed_files.get_targets_from_dataframe(df)\n",
    "    \n",
    "    return targets\n",
    "\n",
    "def create_bed_files(dataset='Lab_Pvalb'):\n",
    "    targets = get_targets()\n",
    "    generate_bed_files.generate_data(targets, dataset, overwrite=True, dataset=dataset)\n",
    "    \n",
    "    return\n",
    "\n",
    "def get_cell_reads(df_data, fname, cell):\n",
    "\n",
    "    df = pd.read_csv(fname, sep='\\t', header=None, names=['Chrome', 'Strand', 'Start', 'End', 'Size'], dtype={'Chrome':str})\n",
    "    df['Chrome'] = df['Chrome'].str.strip('chr')\n",
    "    \n",
    "    for gene, name, chromosome, strand, start, end in df_data.index:\n",
    "        valids = np.logical_and(df.Start<end, df.End>start)\n",
    "        valids = np.logical_and(valids, df['Chrome'] == chromosome)\n",
    "        valids = np.logical_and(valids, df['Strand'] == strand)\n",
    "        df_target = df.loc[valids,:].copy()\n",
    "        df_target.Start = np.clip(df_target.Start, start, None)\n",
    "        df_target.End = np.clip(df_target.End, None, end)\n",
    "        df_target['Span'] = df_target.End - df_target.Start + 1\n",
    "        ratio = df_target.Span.values.astype(float) / df_target.Size.values.astype(float)\n",
    "        df_data.loc[(gene, name, chromosome, strand, start, end), cell] = ratio.sum()\n",
    "    \n",
    "    return\n",
    "\n",
    "def get_read_data(df_data, dataset='Lab_Pvalb'):    \n",
    "    fnames = glob.glob('/media/soma/Storage1/BedFiles/%s/*.bed' % dataset)\n",
    "    fnames.sort()\n",
    "    cells = [fname.split('/')[-1][:-4] for fname in fnames]\n",
    "    \n",
    "    columns = [df_data.Gene, df_data.index, df_data.Chromosome, df_data.Strand, df_data.Start, df_data.End]\n",
    "    names = ('Gene', 'Index', 'Chromosome', 'Strand', 'Start', 'End')\n",
    "    index = pd.MultiIndex.from_arrays(columns, names=names)\n",
    "    \n",
    "    df = pd.DataFrame(np.NaN, index=index, columns=cells)\n",
    "    \n",
    "    for fname, cell in zip(fnames, cells):\n",
    "        get_cell_reads(df, fname, cell)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def adjust_sizes(genes, size):\n",
    "    for gene in np.unique(genes):\n",
    "        ind = np.arange(size.size)[genes==gene]\n",
    "        size[ind[0]] = size[ind[0]] - 35\n",
    "        size[ind[-1]] = max(10, size[ind[-1]]-35)\n",
    "    \n",
    "    return\n",
    "\n",
    "def compile_read_data(dataset='Lab_Pvalb', base_dataset=''):\n",
    "    if len(base_dataset) == 0:\n",
    "        base_dataset = dataset\n",
    "    df_data = get_gene_exons.get_genes_data(get_target_genes())\n",
    "    df = get_read_data(df_data, dataset=dataset)\n",
    "    \n",
    "    fname = '../../reference data/SRRreferences/SRR%s.txt' % base_dataset\n",
    "    df_conv = pd.read_csv(fname, sep='\\t', header=None, skiprows=1, index_col=0, names=['SRR', 'Cell'])\n",
    "    df.columns = df_conv.loc[df.columns,'Cell']\n",
    "    df = df.groupby(df.columns, axis=1).sum()\n",
    "    \n",
    "    size = df.index.get_level_values('End') - df.index.get_level_values('Start') + 1\n",
    "    size = size.values\n",
    "    adjust_sizes(df.index.get_level_values('Gene'), size)\n",
    "    size = size.reshape((size.size,1))\n",
    "    df = df / size * 1000\n",
    "    \n",
    "    df.to_csv('exon defs/exon_rates_%s.tsv' % dataset, sep='\\t')\n",
    "    \n",
    "    return\n",
    "\n",
    "def get_read_data_normalized_by_gene():\n",
    "    df = pd.read_csv('exon defs/exon_rates.tsv', sep='\\t', header=0, index_col=[0,1,2,3,4])\n",
    "    \n",
    "    df_gene = df.groupby(level='Gene', axis=0).sum()\n",
    "    df_gene = df_gene.loc[df.index.get_level_values('Gene'),:]\n",
    "    df_gene.index = df.index\n",
    "    \n",
    "    df = df / df_gene * 100\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_stats(df):\n",
    "    columns = ['Mean', 'Std_Dev', 'Count', 'Std_Err']\n",
    "    df_stat = pd.DataFrame(np.NaN, index=df.index, columns=columns)\n",
    "    df_stat.Mean = df.mean(axis=1)\n",
    "    df_stat.Std_Dev = df.std(axis=1)\n",
    "    df_stat.Count = df.shape[1] - df.isna().sum(axis=1)\n",
    "    df_stat.Std_Err = df_stat.Std_Dev / np.sqrt(df_stat.Count)\n",
    "    \n",
    "    return df_stat\n",
    "\n",
    "def get_plot_data():\n",
    "    df = get_read_data_normalized_by_gene()\n",
    "    df_stat = get_stats(df)\n",
    "    \n",
    "    kept = df_stat.Count > 5\n",
    "    df = df.loc[kept,:].copy()\n",
    "    df_stat = df_stat.loc[kept,:].copy()\n",
    "    \n",
    "    level = df.index.get_level_values\n",
    "    arrays = level('Gene'), level('Index')\n",
    "    names = ('Gene', 'Exon')\n",
    "    index = pd.MultiIndex.from_arrays(arrays, names=names)\n",
    "    df.index = index\n",
    "    df_stat.index = index\n",
    "    \n",
    "    return df, df_stat\n",
    "\n",
    "def plot_gene_map(df, df_stat, gene, ax):\n",
    "    # trim to data of interest\n",
    "    df = df.xs(gene, level='Gene')\n",
    "    df_stat = df_stat.xs(gene, level='Gene')\n",
    "    \n",
    "    # set up axis\n",
    "    ax.set_xticks(np.arange(df_stat.shape[0])+.5)\n",
    "    ax.set_xticklabels(df_stat.index.tolist(), rotation=90, ha='center', fontsize=6)\n",
    "    ax.axis([0, df_stat.shape[0], 0, 100])\n",
    "    ax.set_yticks([0,25,50,75,100])\n",
    "    ax.set_yticklabels([0,25,50,75,100], fontsize=6)\n",
    "    ax.set_title(gene, fontsize=10)\n",
    "    ax.set_ylabel('Expression Rate (%)', fontsize=8)\n",
    "    \n",
    "    # plot statistics\n",
    "    xvals = np.arange(df_stat.shape[0])+.5\n",
    "    ax.plot(xvals, df_stat.Mean, linewidth=1, color='red', zorder=2)\n",
    "    ax.fill_between(xvals, df_stat.Mean-df_stat.Std_Err, df_stat.Mean+df_stat.Std_Err,\n",
    "                    color='#FF6666', alpha=.33, zorder=1, linewidth=0\n",
    "                   )\n",
    "    \n",
    "    # plot individual values\n",
    "    for column, data in df.iteritems():\n",
    "        if data.isna().sum() > 0:\n",
    "            continue\n",
    "        ax.plot(xvals, data, linewidth=.5, color='grey', zorder=0)\n",
    "    \n",
    "    return\n",
    "\n",
    "def plot_gene_maps(df, df_stat):\n",
    "    fig = plt.figure(figsize=(8.5,11))\n",
    "    \n",
    "    genes = np.unique(df.index.get_level_values('Gene'))\n",
    "    \n",
    "    for axnum, gene in enumerate(genes):\n",
    "        col = axnum % 2\n",
    "        row = axnum // 2\n",
    "        \n",
    "        ax = fig.add_axes([.17 + .4*col, .75 - .15 * row, .33, .10])\n",
    "        plot_gene_map(df, df_stat, gene, ax)\n",
    "    \n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = ['GSE60361', 'GSE99888', 'Lab_Pvalb', 'Lab_OLM',\n",
    "            'Cadwell', 'GSE70844', 'GSE119248', 'Gouwens_Hb',\n",
    "            'Tasic_Hb'\n",
    "           ]\n",
    "with open('Times.txt','w') as w:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 14s, sys: 32.2 s, total: 2min 47s\n",
      "Wall time: 32min 32s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "generate_references()\n",
    "for dataset in datasets:\n",
    "    t0 = time.time()\n",
    "    create_bed_files(dataset=dataset)\n",
    "    with open('Times.txt','a') as w:\n",
    "        w.write('Create Bed Files (%s):\\t' % dataset + str(time.time() - t0) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2h 23min 54s, sys: 2.48 s, total: 2h 23min 56s\n",
      "Wall time: 2h 23min 59s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "for dataset in datasets:\n",
    "    t0 = time.time()\n",
    "    compile_read_data(dataset=dataset)\n",
    "    with open('Times.txt','a') as w:\n",
    "        w.write('Compile Read Data (%s):\\t' % dataset + str(time.time() - t0) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1d 6h 27min 7s, sys: 41min 46s, total: 1d 7h 8min 53s\n",
      "Wall time: 1d 7h 9min 57s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "#generate_references()\n",
    "for dataset in datasets:\n",
    "    t0 = time.time()\n",
    "    exon_dist.generate_data(dataset=dataset)\n",
    "    with open('Times.txt','a') as w:\n",
    "        w.write('Generate Distribution (%s):\\t' % dataset + str(time.time() - t0) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
