{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Modules import plot_ax_violin\n",
    "from scipy.stats import ttest_ind\n",
    "from statsmodels.stats import multitest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_ephys_data():\n",
    "    df_ephys, df_seq, df_labels = plot_ax_violin.get_lab_pvalb_dataset()\n",
    "    \n",
    "    columns = [\n",
    "               'Resting membrane potential (mV)', 'Input resistance (MOhm)', 'Capacitance (pF)',\n",
    "               'AP firing threshold (pA)', 'AP symmetricity', 'AP peak amplitude (mV)', \n",
    "               'AP halfwidth (ms)', 'Frequency Slope (Hz/pA)', 'Attenuation',\n",
    "               'Sag potential'\n",
    "              ]\n",
    "    column_names = ['Resting membrane potential (mV)',\n",
    "                    'Input resistance (MOhm)',\n",
    "                    'Capacitance (pF)',\n",
    "                    'AP firing threshold (pA)',\n",
    "                    'AP symmetricity',\n",
    "                    'AP peak amplitude (mV)',\n",
    "                    'AP halfwidth (ms)',\n",
    "                    'Frequency Slope (Hz/pA)',\n",
    "                    'Attenuation',\n",
    "                    'Sag potential (mV)'\n",
    "                   ]\n",
    "    df_ephys = df_ephys.loc[df_ephys.index.isin(df_labels.index),columns].copy()\n",
    "    df_ephys.columns = column_names\n",
    "    df_labels = df_labels.loc[df_ephys.index]\n",
    "    \n",
    "    return df_ephys, df_labels\n",
    "\n",
    "def get_mature_ephys():\n",
    "    df_ephys, df_labels = read_ephys_data()\n",
    "    df_labels = df_labels.loc[df_labels.Age>20]\n",
    "    df_ephys = df_ephys.loc[df_labels.index,:].copy()\n",
    "    df_ephys.index = df_labels['CellType']\n",
    "    \n",
    "    return df_ephys\n",
    "\n",
    "def get_young_old_ephys():\n",
    "    df_ephys, df_labels = read_ephys_data()\n",
    "    df_labels = df_labels.loc[np.logical_or(df_labels.Age<21, df_labels.Age>21)]\n",
    "    df_labels = df_labels.loc[df_labels.CellType=='vBC'].copy()\n",
    "    df_labels['Label'] = ''\n",
    "    young = df_labels.Age<21\n",
    "    mature = df_labels.Age>21\n",
    "    df_labels.loc[young,'Label'] = '<P21'\n",
    "    df_labels.loc[mature,'Label'] = '>P21'\n",
    "    \n",
    "    df_ephys = df_ephys.loc[df_labels.index].copy()\n",
    "    \n",
    "    df_ephys.index = df_labels['Label']\n",
    "    \n",
    "    return df_ephys\n",
    "\n",
    "def get_columns(celltypes, ending):\n",
    "    return ['%s%s' % (celltype, ending) for celltype in celltypes]\n",
    "\n",
    "def summarize_ephys_data(df_ephys):\n",
    "    if 'vAAC' in df_ephys.index:\n",
    "        celltypes = ['vAAC', 'vBIC', 'hBIC', 'vBC', 'hBC']\n",
    "    else:\n",
    "        celltypes = ['<P21', '>P21']\n",
    "    endings = [' (Count)', ' (Mean)', ' (Standard Deviation)', ' (Minimum)', ' (Maximum)']\n",
    "    columns = [column for ending in endings for column in get_columns(celltypes, ending)]\n",
    "    if len(celltypes) == 2:\n",
    "        columns = ['t-score', 'P-value', 'FDR'] + columns\n",
    "    df_stat = pd.DataFrame(np.NaN, index=df_ephys.columns, columns=columns)\n",
    "    \n",
    "    df_group = df_ephys.T.groupby(df_ephys.index, axis=1)\n",
    "    df_stat.loc[:,get_columns(celltypes, ' (Count)')] = df_group.count()[celltypes].values\n",
    "    df_stat.loc[:,get_columns(celltypes, ' (Mean)')] = df_group.mean()[celltypes].values\n",
    "    df_stat.loc[:,get_columns(celltypes, ' (Standard Deviation)')] = df_group.std()[celltypes].values\n",
    "    df_stat.loc[:,get_columns(celltypes, ' (Minimum)')] = df_group.min()[celltypes].values\n",
    "    df_stat.loc[:,get_columns(celltypes, ' (Maximum)')] = df_group.max()[celltypes].values\n",
    "    for column in get_columns(celltypes, ' (Count)'):\n",
    "        df_stat[column] = df_stat[column].astype(int)\n",
    "    \n",
    "    if len(celltypes) == 2:\n",
    "        df_1 = df_ephys.loc[celltypes[0]]\n",
    "        df_2 = df_ephys.loc[celltypes[1]]\n",
    "        t_vals, pvals = ttest_ind(df_1.values, df_2.values, axis=0, equal_var=False, nan_policy='omit')\n",
    "        df_stat['t-score'] = t_vals\n",
    "        df_stat['P-value'] = pvals\n",
    "        df_stat['FDR'] = multitest.multipletests(pvals, method='fdr_bh')[1]\n",
    "    \n",
    "    return df_stat\n",
    "\n",
    "def get_differential_genes(target, fdr=.15):\n",
    "    target = target.split('-PV-types')[0]\n",
    "    directory = 'Differential/edgeR/%s' % target\n",
    "    fnames = glob.glob('%s/*.txt' % directory)\n",
    "    fnames.sort()\n",
    "    \n",
    "    data_frames = []\n",
    "    \n",
    "    for fname in fnames:\n",
    "        df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "        df = df.loc[np.abs(df.logFC)>1]\n",
    "        df_sig = df.loc[df.FDR<fdr].copy()\n",
    "        compare = fname.split('/')[-1][:-4].split(' vs ')\n",
    "\n",
    "        \n",
    "        if df_sig.shape[0] > 0:\n",
    "            df_sig['Category 1'] = compare[0]\n",
    "            df_sig['Category 2'] = compare[1]\n",
    "            data_frames.append(df_sig)\n",
    "        \n",
    "    return pd.concat(data_frames, axis=0)\n",
    "\n",
    "def get_differential_stats(target, count=8):\n",
    "    directory = 'Differential/edgeR/%s' % target\n",
    "    fnames = glob.glob('%s/*.txt' % directory)\n",
    "    fnames.sort()\n",
    "    \n",
    "    data_frames = []\n",
    "    \n",
    "    for fname in fnames:\n",
    "        df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "        data_frames.append(df.loc[:,['logFC', 'FDR']])\n",
    "    \n",
    "    names = [fname.split('/')[-1][:-4] for fname in fnames]\n",
    "    df_fc = pd.concat([df[['logFC']] for df in data_frames], axis=1, join='outer', sort=False)\n",
    "    df_fdr = pd.concat([df[['FDR']] for df in data_frames], axis=1, join='outer', sort=False)\n",
    "    df_sig = df_fdr.copy()\n",
    "    df_sig[np.abs(df_fc)<=1] = np.NaN\n",
    "    gene_signifs = df_sig.min(axis=1)\n",
    "    genes = gene_signifs.nsmallest(n=count).index.tolist()\n",
    "    \n",
    "    df_fc = df_fc.loc[genes,:]\n",
    "    df_fdr = df_fdr.loc[genes,:]\n",
    "    \n",
    "    df_fc.columns = names\n",
    "    df_fdr.columns = names\n",
    "        \n",
    "    return df_fc, df_fdr\n",
    "\n",
    "def get_differential_data(targets, fdr=.15):\n",
    "    df = pd.concat([get_differential_genes(target, fdr=fdr) for target in targets], axis=0)\n",
    "    df = df.loc[:,['logFC', 'PValue', 'FDR', 'Category 1', 'Category 2']].copy()\n",
    "    df.columns = ['log2 Fold Change', 'p-value', 'p-adjusted (FDR)', 'Category 1', 'Category 2']\n",
    "    \n",
    "    params = {'sep':'\\t', 'header':0, 'index_col':0}\n",
    "    df_rate = pd.read_csv('Mapping/DifferentialMapping/Lin_Harris_rates.tsv', **params)\n",
    "    df_rate = df_rate.loc[df_rate.max(axis=1)>0,:]\n",
    "    df['Rate in this study (%)'] = df_rate.loc[df.index,'This Study']\n",
    "    df['Rate in CA1-INs study (%)'] = df_rate.loc[df.index,'CA1-INs']\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_sholl_data():\n",
    "    fname = 'References/Sholl_Soma.tsv'\n",
    "    df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    df.sort_values('Age', inplace=True)\n",
    "    df.columns = [column.replace(' um', ' Î¼m') for column in df.columns]\n",
    "    df['Sholl Analysis on dendritic length;'] = ''\n",
    "    df = df.iloc[:,[0,2,1,-1] + list(i for i in range(3,df.shape[1]-1))]\n",
    "    \n",
    "    return df\n",
    "\n",
    "def get_summary():\n",
    "    index = ['Figure 1C',\n",
    "             'Figure 1D_1',\n",
    "             'Figure 1D_2',\n",
    "             'Figure 1E',\n",
    "             'Figure 2A',\n",
    "             'Figure 2C',\n",
    "             'Figure 2E',\n",
    "             'Figure 3A',\n",
    "             'Figure 3B',\n",
    "             'Figure 4A-E',\n",
    "             'Figure 5B_1',\n",
    "             'Figure 5B_2',\n",
    "             'Figure 5B_3',\n",
    "             'Figure 5C_1',\n",
    "             'Figure 5C_2',\n",
    "             'Figure 5C_3',\n",
    "             'Figure 5D',\n",
    "             'Figure 6B-D',\n",
    "             'Figure 6E',\n",
    "             'Figure 6F',\n",
    "             'Figure 7A',\n",
    "             'Figure 7C',\n",
    "            ]\n",
    "    columns = ['Figure', 'Figure Title', 'Plot', 'Plot Title', 'Summary']\n",
    "    df = pd.DataFrame('', index=index, columns=columns)\n",
    "    df.index.name = 'Sheet Names'\n",
    "    \n",
    "    figure_titles = {'Figure 1':'Morpho-transcriptomic profiling of PV-INs',\n",
    "                     'Figure 2':'Transcriptomic properties of PV-INs',\n",
    "                     'Figure 3':'Electrophysiological properties of PV-INs',\n",
    "                     'Figure 4':'Transcriptomic correlates of morphological PV types',\n",
    "                     'Figure 5':'Analysis of CAM expression in morphological PV types',\n",
    "                     'Figure 6':'Morphological and electrophysiological analysis of vBC'\\\n",
    "                                ' type PV-INs during circuit maturation',\n",
    "                     'Figure 7':'Age-dependent transcriptomic changes and'\\\n",
    "                                ' onset of hemoglobin expression in PV-INs'\n",
    "                    }\n",
    "    summaries = {'1C':  'Statistical information on differentially expressed genes between'\\\n",
    "                        ' PV-INs and SST-OLMs. Columns show log2 of fold change, p-value for'\\\n",
    "                        ' significance, p-adjusted (FDR) for multiple testing adjusted'\\\n",
    "                        ' significance, and whether a gene was enriched in PV-INs or SST-OLMs.'\\\n",
    "                        ' Statistics were calculated with edgeR.',\n",
    "                 '1D_1':'Cell names, morphological types, and UMAP embedding coordinates for'\\\n",
    "                        ' PV-INs and SST-OLM cells.',\n",
    "                 '1D_2':'Cell names, morphological types, and UMAP embedding coordinates for'\\\n",
    "                        ' PV-INs.',\n",
    "                 '1E':  'Average classification accuracy, and standard deviation for Random Forest'\\\n",
    "                        ' classifications of PV-INs vs OLMs, BCs vs non-BCs, and BICs vs non-BICs.',\n",
    "                 '2A':  'Cell names, morphological type, proMMT generated cell types, and'\\\n",
    "                        ' nbt-SNE generated coordinates for PV-INs.',\n",
    "                 '2C':  'List of 25 genes that are differentially expressed at a cutoff of p-adjusted<0.05'\\\n",
    "                        ' between at least 1 pair of proMMT types. If a gene is DE across multiple comparisons,'\\\n",
    "                        ' it is listed multiple times. Statistics are calculated with edgeR. For each gene, we'\\\n",
    "                        ' list the log2 fold change, p-value, p-adjusted, the two categories across which the'\\\n",
    "                        ' comparison happened, and the % of cells that expressed the gene in our PV-INs, and the'\\\n",
    "                        ' PV-INs from the CA1-INs study.',\n",
    "                 '2E':  'Quantification of percentage of PV-INs that mapped to a given continent'\\\n",
    "                        ' in the CA1-INs data, using different sets of reference genes.',\n",
    "                 '3A':  'Electrophysiological properties of different morphological types.'\\\n",
    "                        ' For each morphological type, we look at number of cells (count), average'\\\n",
    "                        ' expression (mean), standard deviation, minimum, and maximum.',\n",
    "                 '3B':  'UMAP based plotting of data of PV-INs using electrophysiological properties.'\n",
    "                        ' For each cell we display, we have their dendro-morphological type,'\\\n",
    "                        ' axo-morphological type, proMMT generated type, morphological type, and UMAP'\\\n",
    "                        ' embedding coordinates.',\n",
    "                 '4A-E':'List of 52 genes that are differentially expressed at a cutoff of p-adjusted<.15'\\\n",
    "                        ' between either morphological types, dendro-morphological types, or '\\\n",
    "                        ' axo-morphological types. If a gene is DE across multiple comparisons, it is listed '\\\n",
    "                        ' multiple times. Statistics are calculated with edgeR. For each gene, we list the'\\\n",
    "                        ' log2 fold change, p-value, p-adjusted, the two categories'\\\n",
    "                        ' across which the comparison happened, and the % of cells that expressed'\\\n",
    "                        ' the gene in our PV-INs, and the PV-INs from the CA1-INs study.',\n",
    "                 '5B_1':'Log2 Fold changes of most significant CAMs in pairwise PV-IN subtype'\\\n",
    "                        ' comparisons. In cases where a gene was not expressed in at least 5 cells'\\\n",
    "                        ' with a level of TPM>15 between the two groups in the comparison, comparisons'\\\n",
    "                        ' were not made, and the values were left blank.',\n",
    "                 '5B_2':'FDRs of most significant CAMs in pairwise PV-IN subtype comparisons.'\\\n",
    "                        ' Comparisons. In cases where a gene was not expressed in at least 5 cells'\\\n",
    "                        ' with a level of TPM>15 between the two groups in the comparison, comparisons'\\\n",
    "                        ' were not made, and the values were left blank.',\n",
    "                 '5B_3':'Log2 Fold changes and FDRs of top 8 most differentially expressed genes'\\\n",
    "                        ' between PV-INs and SST-OLM cells.',\n",
    "                 '5C_1':'Similarity between morphological PV types and SST-OLMs.'\\\n",
    "                        ' Similarities between cells were calculated as Pearson correlation of log'\\\n",
    "                        ' of CAM TPM expressions. To calculate similarities between two cell types,'\\\n",
    "                        ' the average of all pairwise similarities of cells in the two cell types were'\\\n",
    "                        ' taken. To not bias the results, when calculating the similarity of a cell type'\\\n",
    "                        ' with itself, the distance of cells to themselves were not used in the calculation.',\n",
    "                 '5C_2':'Similarity between axo-morphological PV types and SST-OLMs.'\\\n",
    "                        ' Similarities between cells were calculated as Pearson correlation of log'\\\n",
    "                        ' of CAM TPM expressions. To calculate similarities between two cell types,'\\\n",
    "                        ' the average of all pairwise similarities of cells in the two cell types were'\\\n",
    "                        ' taken. To not bias the results, when calculating the similarity of a cell type'\\\n",
    "                        ' with itself, the distance of cells to themselves were not used in the calculation.',\n",
    "                 '5C_3':'Similarity between dendro-morphological PV-IN types and SST-OLMs.'\\\n",
    "                        ' Similarities between cells were calculated as Pearson correlation of log'\\\n",
    "                        ' of CAM TPM expressions. To calculate similarities between two cell types,'\\\n",
    "                        ' the average of all pairwise similarities of cells in the two cell types were'\\\n",
    "                        ' taken. To not bias the results, when calculating the similarity of a cell type'\\\n",
    "                        ' with itself, the distance of cells to themselves were not used in the calculation.',\n",
    "                 '5D':  'Log 2 of expression levels of specific CAMs in PV-INs. CAMs are taken from the'\\\n",
    "                        ' FÃ¶ldy et al. (2016), and were shown as markers for regular-spiking,'\\\n",
    "                        ' fast-spiking, or CA1 Pyramidal cells. Our PV-INs have expression consistent'\\\n",
    "                        ' with what they showed for fast-spiking cells. Rows are genes, and columns are'\\\n",
    "                        ' cells with their morphological cell types below the cell names.',\n",
    "                 '6B-D':'Morphological data for a subset of vBCs that were used in our'\\\n",
    "                        ' morphological analysis. For each cell we list its name, age, the length'\\\n",
    "                        ' of its dendrite, the length of its axon, and the number of intersections at'\\\n",
    "                        ' 25Î¼m intervals.',\n",
    "                 '6E':  'Electrophysiological comparison of <P21 and >P21 vBCs. For each age'\\\n",
    "                        '  category we have number of cells (Count), average expression'\\\n",
    "                        ' (Mean), standard deviation, minimum, and maximum for all 10 electrophysiological'\\\n",
    "                        ' properties.',\n",
    "                 '6F':  'UMAP based embedding of vertical Basket Cells using electrophysiological properties.'\\\n",
    "                        ' Columns are cell name, cell age (in days), x-coordinate, and y-coordinate of UMAP'\\\n",
    "                        ' embedding.',\n",
    "                 '7A':  'Genes that were found to be up-regulated or down-regulated with age in vBCs,'\\\n",
    "                        ' using the Gini impurity measure. Statistics were calculated using'\\\n",
    "                        ' a Monte Carlo method. Switching Age lists the cut-off point where the sliding'\\\n",
    "                        ' window gives the best Gini impurity difference. p-value and p-adjusted are'\\\n",
    "                        ' measures of significance, with and without accounting for multiple testing.'\\\n",
    "                        ' A gene is considered up-regulated if it is \"turned on\" with age, and down-regulated'\\\n",
    "                        ' if it is \"turned off\" after the switching age. Cutoff for significance was'\\\n",
    "                        ' p-adjusted < 0.10. As such, about 5 or 6 of the 53 genes listed here are'\\\n",
    "                        ' expected to not be actually switching genes.',\n",
    "                 '7C':  'Measure of accuracy in predicting the age of a cell using a random forest'\\\n",
    "                        ' classifier. For each of our PV-INs, we show cell name, morphological type,'\\\n",
    "                        ' cell age, and the rate (%) with which it is predicted to be older than P25.'\\\n",
    "                        ' Only vBCs were using in training the classifier.'\n",
    "                }\n",
    "    \n",
    "    df['Figure'] = [ind[:8] for ind in df.index]\n",
    "    df['Figure Title'] = df['Figure'].map(figure_titles)\n",
    "    df['Plot'] = [ind[7:].split('_')[0] for ind in df.index]\n",
    "    df['Plot Title'] = [ind[7:] for ind in df.index]\n",
    "    df['Summary'] = df['Plot Title'].map(summaries)\n",
    "    \n",
    "    df = df.loc[:,['Figure', 'Figure Title', 'Plot', 'Summary']]\n",
    "    \n",
    "    return df\n",
    "\n",
    "def set_width(worksheet, index, name, column, column_format):\n",
    "    max_len = max((\n",
    "        column.astype(str).map(len).max(),\n",
    "        len(str(column.name))\n",
    "        ))\n",
    "    \n",
    "    if type(name) is tuple:\n",
    "        max_len = max(max_len, max([len(nam) for nam in name])) + 1\n",
    "    elif type(name) is str:\n",
    "        max_len = max(max_len, len(name)+1) + 1\n",
    "    else:\n",
    "        max_len += 1\n",
    "    \n",
    "    worksheet.set_column(index, index, max_len, column_format)\n",
    "    \n",
    "    return\n",
    "\n",
    "def adjust_dataframe_numbers(df):\n",
    "    for col, data in df.items():\n",
    "        if data.dtype == np.float64:\n",
    "            if np.abs(data).max()>1:\n",
    "                df[col] = np.around(data,2)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_excel(writer, df, sheet_name, writer_args, wrap_text=False):\n",
    "    adjust_dataframe_numbers(df)\n",
    "    df.to_excel(writer, sheet_name=sheet_name)\n",
    "    worksheet = writer.sheets[sheet_name]\n",
    "    set_width(worksheet, 0, df.index.name, df.index, writer_args['format'])\n",
    "    for idx, (col, data) in enumerate(df.items(), start=1):\n",
    "        if wrap_text and col == df.columns[-1]:\n",
    "            continue\n",
    "        set_width(worksheet, idx, col, data, writer_args['format'])\n",
    "    \n",
    "    if wrap_text:\n",
    "        column = df.columns[-1]\n",
    "        worksheet.set_column(df.shape[1], df.shape[1], 100, writer_args['format'])\n",
    "    \n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define Excel Generating Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_summary(writer, writer_args):\n",
    "    df = get_summary()\n",
    "    \n",
    "    write_excel(writer, df, 'Summary', writer_args, wrap_text=True)\n",
    "    \n",
    "    return\n",
    "    \n",
    "def write_1C(writer, writer_args):\n",
    "    fname = 'Differential/edgeR/General/Sst-OLM vs Pvalb.txt'\n",
    "    df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    df = df.loc[np.logical_and(np.abs(df.logFC)>1, df.FDR<.05),['logFC', 'PValue', 'FDR']].copy()\n",
    "    df['Regulated'] = (df.logFC>0).map({True:'PV', False:'SST'})\n",
    "    df.columns = ['log2 Fold Change', 'p-value', 'p-adjusted (FDR)', 'Enriched In']\n",
    "    df.sort_values('p-adjusted (FDR)', inplace=True)\n",
    "    \n",
    "    write_excel(writer, df, 'Figure 1C', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_1D(writer, writer_args):\n",
    "    fname = 'Mapping/Embeddings/Lab_Pvalb+Lab_OLM.tsv'\n",
    "    df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    write_excel(writer, df, 'Figure 1D_1', writer_args)\n",
    "    \n",
    "    fname = 'Mapping/Embeddings/Lab_Pvalb_Only.tsv'\n",
    "    df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    write_excel(writer, df, 'Figure 1D_2', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_1E(writer, writer_args):\n",
    "    fname = 'Mapping/Separation Accuracy.tsv'\n",
    "    df = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    write_excel(writer, df, 'Figure 1E', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_2A(writer, writer_args):\n",
    "    fname = 'Mapping/Embeddings/Lab_Pvalb-tpm.tsv'\n",
    "    params = {'sep':'\\t', 'header':0, 'index_col':0}\n",
    "    df_embedding_tpm = pd.read_csv(fname, **params)\n",
    "\n",
    "    fname = 'Datasets/Lab_Pvalb-transcriptional_labels.tsv'\n",
    "    df_labels = pd.read_csv(fname, na_values='Other', **params)\n",
    "    df_labels = df_labels.loc[df_labels.Age>20]\n",
    "    df_embedding_tpm = df_embedding_tpm.loc[df_labels.index,:]\n",
    "\n",
    "    df = pd.concat((df_labels.loc[:,['Morph-PV-types', 'proMMT-PV-types']],\n",
    "                    df_embedding_tpm.loc[:,['nbt-SNE_X', 'nbt-SNE_Y']]),\n",
    "                   axis=1)\n",
    "    df.columns = ['Morphology', 'proMMT classification', 'nbt-SNE X-coordinate', 'nbt-SNE Y-coordinate']\n",
    "    \n",
    "    write_excel(writer, df, 'Figure 2A', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_2C(writer, writer_args):\n",
    "    targets = ['proMMT']\n",
    "    df = get_differential_data(targets, fdr=0.05)\n",
    "    df.sort_values('p-adjusted (FDR)', inplace=True)\n",
    "    write_excel(writer, df, 'Figure 2C', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_2E(writer, writer_args):\n",
    "    fname = 'Mapping/Excel/Map_to_Cell_Types.xlsx'\n",
    "    df = pd.read_excel(fname, sheet_name='Cell Mappings (knn = 25)', index_col=[0,1]).fillna(0).astype(int)\n",
    "    \n",
    "    fname = 'Datasets/Lab_Pvalb-labels.tsv'\n",
    "    df_labels = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    df_labels = df_labels.loc[df_labels.Age>20]\n",
    "    df = df.loc[df.index.get_level_values('Cell').isin(df_labels.index)]\n",
    "    \n",
    "    df = pd.concat([data.value_counts() for col, data in df.iteritems()], axis=1, join='outer').fillna(0).astype(int)\n",
    "    df.index = df.index.map({2:'Pvalb.Tac1', 3:'Pvalb.C1ql1'}).fillna('Other')\n",
    "    df_counts = df.groupby(df.index).sum()\n",
    "    \n",
    "    df_counts = df_counts / df_counts.sum(axis=0) * 100\n",
    "    columns = ['Chi_Squared',\n",
    "               'Correlation',\n",
    "               'F_Classif',\n",
    "               'Harris_Genes',\n",
    "               'Mutual_Info',\n",
    "               'PV_Pvalb_Kobak_Lin'\n",
    "              ]\n",
    "    column_names = ['Chi-squared',\n",
    "                    'Correlation',\n",
    "                    'F-classification',\n",
    "                    'Harris et al.',\n",
    "                    'Mutual information',\n",
    "                    'Kobak on Pvalb'\n",
    "                   ]\n",
    "    df_counts = df_counts.loc[:,columns].copy()\n",
    "    df_counts.columns=column_names\n",
    "    \n",
    "    write_excel(writer, df_counts, 'Figure 2E', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_3A(writer, writer_args):\n",
    "    df_ephys = get_mature_ephys()\n",
    "    df_stat = summarize_ephys_data(df_ephys)\n",
    "    \n",
    "    write_excel(writer, df_stat, 'Figure 3A', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_3B(writer, writer_args):\n",
    "    df_embed = pd.read_csv('Mapping/Embeddings/Lab_Pvalb-ephys.tsv', sep='\\t', header=0, index_col=0)\n",
    "    fname = 'References/marker_ref.txt'\n",
    "    \n",
    "    fname = 'Datasets/Lab_Pvalb-transcriptional_labels.tsv'\n",
    "    kwargs = {'sep':'\\t', 'header':0, 'index_col':0}\n",
    "    df_labels = pd.read_csv(fname, **kwargs)\n",
    "    columns = ['MorphDirectional-PV-types', 'MorphMarker-PV-types', 'proMMT-PV-types', 'Morph-PV-types']\n",
    "    column_names = ['Dendro-morphological types',\n",
    "                    'Axo-morphological types',\n",
    "                    'proMMT types',\n",
    "                    '5 morphological types',\n",
    "                   ]\n",
    "    df = df_labels.loc[df_embed.index, columns].copy()\n",
    "    df.columns = column_names\n",
    "    df['UMAP_X'] = df_embed['UMAP_X']\n",
    "    df['UMAP_Y'] = df_embed['UMAP_Y']\n",
    "    \n",
    "    write_excel(writer, df, 'Figure 3B', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_4AE(writer, writer_args):\n",
    "    targets = ['Morph-PV-types',\n",
    "               'MorphMarker-PV-types',\n",
    "               'MorphDirectional-PV-types'\n",
    "              ]\n",
    "    df = get_differential_data(targets, fdr=.15)\n",
    "    df.sort_values('p-adjusted (FDR)', inplace=True)\n",
    "    write_excel(writer, df, 'Figure 4A-E', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_5B(writer, writer_args):\n",
    "    df_fc, df_fdr = get_differential_stats('Morph_CAM', count=8)\n",
    "    write_excel(writer, df_fc, 'Figure 5B_1', writer_args)\n",
    "    write_excel(writer, df_fdr, 'Figure 5B_2', writer_args)\n",
    "    \n",
    "    df_fc, df_fdr = get_differential_stats('Lab_CAM', count=8)\n",
    "    df = pd.DataFrame(np.NaN, index=df_fc.index, columns=['log2FC', 'FDR'])\n",
    "    df.FDR = df_fdr.iloc[:,0]\n",
    "    df.log2FC = df_fc.iloc[:,0]\n",
    "    write_excel(writer, df, 'Figure 5B_3', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_5C(writer, writer_args):\n",
    "    kwargs = {'sep':'\\t', 'header':0, 'index_col':0}\n",
    "    file_names = ['Morphology', 'Axomorphic', 'Dendromorphic']\n",
    "    sheet_names = ['Figure 5C_1', 'Figure 5C_2', 'Figure 5C_3']\n",
    "    \n",
    "    for file_name, sheet_name in zip(file_names, sheet_names):\n",
    "        df = pd.read_csv('CAM Data/%s.tsv' % file_name, **kwargs)\n",
    "        write_excel(writer, df, sheet_name, writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_5D(writer, writer_args):\n",
    "    kwargs = {'sep':'\\t', 'header':[0,1], 'index_col':0}\n",
    "    fname = 'CAM Data/PNAS heatmap.tsv'\n",
    "    df = pd.read_csv(fname, **kwargs)\n",
    "    \n",
    "    order = ['vAAC', ' ', 'vBIC', ' ', 'hBIC', ' ', 'vBC', ' ', 'hBC']\n",
    "    celltypes = df.columns.get_level_values('CellType')\n",
    "    cells = [cell for celltype in order for cell in df.columns[celltypes==celltype]]\n",
    "    df = df.loc[:,cells].copy()\n",
    "    \n",
    "    write_excel(writer, df, 'Figure 5D', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_6BD(writer, writer_args):\n",
    "    df = get_sholl_data()\n",
    "    \n",
    "    write_excel(writer, df, 'Figure 6B-D', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_6E(writer, writer_args):\n",
    "    df_ephys = get_young_old_ephys()\n",
    "    df_stat = summarize_ephys_data(df_ephys)\n",
    "    \n",
    "    write_excel(writer, df_stat, 'Figure 6E', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_6F(writer, writer_args):\n",
    "    df = pd.read_csv('Mapping/Embeddings/Lab_Pvalb_vBC-electro.tsv', sep='\\t', header=0, index_col=0)\n",
    "    \n",
    "    write_excel(writer, df, 'Figure 6F', writer_args)\n",
    "    \n",
    "    return\n",
    "\n",
    "def write_7A(writer, writer_args):\n",
    "    fname = 'Gini Monte Carlo/Gene Scores.tsv'\n",
    "    df_scores = pd.read_csv(fname, sep='\\t', header=0, index_col=0)\n",
    "    df_scores = df_scores.loc[df_scores.P_Adj <= .1].copy()\n",
    "    df_scores.sort_values(['Up_Down', 'Cutoff_Age'], ascending=False, inplace=True)\n",
    "    df_scores = df_scores.loc[:,['Cutoff_Age', 'P_Value', 'P_Adj', 'Up_Down']].copy()\n",
    "    df_scores.columns = ['Switching Age', 'p-value', 'p-adjusted (FDR)', 'Up or down regulated']\n",
    "    \n",
    "    write_excel(writer, df_scores, 'Figure 7A', writer_args)\n",
    "    return\n",
    "\n",
    "def write_7C(writer, writer_args):\n",
    "    kwargs = {'sep':'\\t', 'header':0, 'index_col':0}\n",
    "    fname = 'Datasets/Lab_Pvalb-transcriptional_labels.tsv'\n",
    "    df_labels = pd.read_csv(fname, **kwargs)\n",
    "    \n",
    "    fname = 'Datasets/Lab_Pvalb-Age-Predictions.tsv'\n",
    "    df_predict = pd.read_csv(fname, **kwargs)\n",
    "    df_predict = df_predict.loc[df_labels.index,:].iloc[:,1:].copy()\n",
    "    \n",
    "    columns = ['CellType', 'Age', '>P25 Prediction Rate (%)']\n",
    "    df_summary = pd.DataFrame(np.NaN, index=df_predict.index, columns=columns)\n",
    "    df_summary['CellType'] = df_labels['Morph-PV-types']\n",
    "    df_summary['Age'] = df_labels['Age']\n",
    "    df_summary['>P25 Prediction Rate (%)'] = df_predict.mean(axis=1) * 100\n",
    "    df_summary.sort_values('Age', inplace=True)\n",
    "    \n",
    "    write_excel(writer, df_summary, 'Figure 7C', writer_args)\n",
    "    return\n",
    "\n",
    "def write_figure(writer, writer_args, fignum):\n",
    "    ref_funcs = {1:(write_1C, write_1D, write_1E),\n",
    "                 2:(write_2A, write_2C, write_2E),\n",
    "                 3:(write_3A, write_3B),\n",
    "                 4:(write_4AE,),\n",
    "                 5:(write_5B, write_5C, write_5D),\n",
    "                 6:(write_6BD, write_6E, write_6F),\n",
    "                 7:(write_7A, write_7C)\n",
    "                }\n",
    "    \n",
    "    functions = ref_funcs[fignum]\n",
    "    \n",
    "    for function in functions:\n",
    "        function(writer, writer_args)\n",
    "        \n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/soma/anaconda3/envs/analyze/lib/python3.6/site-packages/ipykernel_launcher.py:7: PerformanceWarning: indexing past lexsort depth may impact performance.\n",
      "  import sys\n",
      "/home/soma/anaconda3/envs/analyze/lib/python3.6/site-packages/ipykernel_launcher.py:244: PerformanceWarning: indexing past lexsort depth may impact performance.\n"
     ]
    }
   ],
   "source": [
    "with pd.ExcelWriter('Manuscripts/Que_et_al.xlsx', engine='xlsxwriter') as writer:\n",
    "    workbook=writer.book\n",
    "    base_format = workbook.add_format({'text_wrap': True, 'valign': 'top'})\n",
    "    kwargs = {'format':base_format}\n",
    "    write_summary(writer, kwargs)\n",
    "    for figure in (1,2,3,4,5,6,7):\n",
    "        write_figure(writer, kwargs, figure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
